src/isl/util.rs

```diff
@@ -0,0 +1,286 @@
+use crate::result::{invalid_schema_error, IonSchemaResult};
+use ion_rs::types::decimal::Decimal;
+use ion_rs::types::timestamp::Timestamp;
+use ion_rs::value::owned::{text_token, OwnedElement, OwnedSymbolToken};
+use ion_rs::value::{AnyInt, Element, IntAccess, Sequence, SymbolToken};
+use ion_rs::IonType;
+use num_traits::Signed;
+use std::convert::TryInto;
+
+/// Represents ISL [Range]s where some constraints can be defined by a range
+/// <RANGE<RANGE_TYPE>> ::= range::[ <EXCLUSIVITY><RANGE_TYPE>, <EXCLUSIVITY><RANGE_TYPE> ]
+///                       | range::[ min, <EXCLUSIVITY><RANGE_TYPE> ]
+///                       | range::[ <EXCLUSIVITY><RANGE_TYPE>, max ]
+/// Grammar: <RANGE_TYPE> ::= <DECIMAL>
+///                | <FLOAT>
+///                | <INT>
+///                | <NUMBER>
+///                | <TIMESTAMP>
+///                | <TIMESTAMP_PRECISION_VALUE>
+/// For more information on [Range]: <https://amzn.github.io/ion-schema/docs/spec.html#constraints>
+#[derive(Debug, Clone, PartialEq)]
+pub enum Range {
+    Decimal(RangeBoundaryValue, RangeBoundaryValue),
+    Float(RangeBoundaryValue, RangeBoundaryValue),
+    Integer(RangeBoundaryValue, RangeBoundaryValue),
+    IntegerNonNegative(RangeBoundaryValue, RangeBoundaryValue),
+    Timestamp(RangeBoundaryValue, RangeBoundaryValue),
+}
+
+impl Range {
+    pub fn decimal_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Decimal(start, end))
+    }
+
+    pub fn float_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Float(start, end))
+    }
+
+    pub fn int_range(start: RangeBoundaryValue, end: RangeBoundaryValue) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Integer(start, end))
+    }
+
+    pub fn int_non_negative_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::IntegerNonNegative(start, end))
+    }
+
+    pub fn timestamp_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Timestamp(start, end))
+    }
+
+    pub fn from_ion_element(value: &OwnedElement, is_non_negative: bool) -> IonSchemaResult<Range> {
```

Comment:
The `is_non_negative` flag feels a bit unfortunate to me. It's only relevant in the integer case to distinguish between `Integer` and `IntegerNonNegative` but has to be specified no matter which type is expected. It also has to be passed through to each of the nested calls to other `from_ion_element` implementations.

Looking at [the grammar](https://amzn.github.io/ion-schema/docs/spec.html#grammar), it seems like `RANGE_TYPE` doesn't include a `NonNegativeInteger` variant: 
```
<RANGE_TYPE> ::= <DECIMAL>
               | <FLOAT>
               | <INT>
               | <NUMBER>
               | <TIMESTAMP>
               | <TIMESTAMP_PRECISION_VALUE>
```

Instead, it [creates a type](https://amzn.github.io/ion-schema/docs/spec.html#byte_length-constraint) to model this constraint:

```
type::{
  name: range_int_non_negative,
  type: list,
  annotations: required::[range],
  ordered_elements: [
    { one_of: [ int_non_negative, { valid_values: [min] } ] },
    { one_of: [ int_non_negative, { valid_values: [max] } ] },
  ],
  container_length: 2,
}
```

I'm inclined to say we should follow suit here. We would remove the `NonNegativeInteger` variant from the `RangeBoundaryValueType` and model this as a type instead. `from_ion_element` would then be able to infer everything it needs from the Ion data itself.

Reply:
I have created an issue for this to change it after I have ordered_elements implemented as a constraint. (Reference issue: https://github.com/amzn/ion-schema-rust/issues/46)